src/cutadapt/_kmer_finder.pyx

```diff
@@ -0,0 +1,303 @@
+            set_masks(needle_mask, i, "GgCc")
+            if query_wildcards:
+                set_masks(needle_mask, i, "YRSKMBDHVNyrskmbdhvn")
+        elif (c == b"W" or c == b"w") and ref_wildcards:
+            set_masks(needle_mask, i, "AaTtUu")
+            if query_wildcards:
+                set_masks(needle_mask, i, "YRWKMBDHVNyrwkmbdhvn")
+        elif (c == b"K" or c == b"k") and ref_wildcards:
+            set_masks(needle_mask, i, "GgTtUu")
+            if query_wildcards:
+                set_masks(needle_mask, i, "YRWSKBDHVNyrwskbdhvn")
+        elif (c == b"M" or c == b"m") and ref_wildcards:
+            set_masks(needle_mask, i, "AaCc")
+            if query_wildcards:
+                set_masks(needle_mask, i, "YRWSMBDHVNyrwsmbdhvn")
+        elif (c == b"B" or  c == b"b") and ref_wildcards:
+            set_masks(needle_mask, i, "CcGgTtUu")
+            if query_wildcards:
+                set_masks(needle_mask, i, "RYSWKMBDHVNryswkmbdhvn")
+        elif (c == b"D" or c == b"d") and ref_wildcards:
+            set_masks(needle_mask, i, "AaGgTtUu")
+            if query_wildcards:
+                set_masks(needle_mask, i, "RYSWKMBDHVNryswkmbdhvn")
+        elif (c == b"H" or c == b"h") and ref_wildcards:
+            set_masks(needle_mask, i, "AaCcTtUu")
+            if query_wildcards:
+                set_masks(needle_mask, i, "RYSWKMBDHVNryswkmbdhvn")
+        elif (c == b"V" or c == b"v") and ref_wildcards:
+            set_masks(needle_mask, i, "AaCcGg")
+            if query_wildcards:
+                set_masks(needle_mask, i, "RYSWKMBDHVNryswkmbdhvn")
+        elif (c == b"N" or c == b"n") and ref_wildcards:
+            if query_wildcards:  # Proper IUPAC matching
+                set_masks(needle_mask, i, "ACGTURYSWKMBDHVNacgturyswkmbdhvn")
+            else:  # N matches literally everything except \00
+                for j in range(1,128):
+                    needle_mask[j] &= ~(<bitmask_t>1ULL << i)
+        elif query_wildcards and not ref_wildcards:
+            # All non-AGCT characters match to N
+            set_masks(needle_mask, i, "Nn")
+        elif ref_wildcards:
+            # ref and query wildcards are True. Perform proper IUPAC matching.
+            # All unknown characters do not match.
+            pass
+        else:
+            if chr(c).isalpha():
+                bothcase = chr(c).lower() + chr(c).upper()
+                set_masks(needle_mask, i, bothcase.encode("ascii"))
+            else:
+                needle_mask[<uint8_t>c] &= ~(<bitmask_t>1ULL << i)
+
+
+cdef bint shift_or_multiple_is_present(
+    const char *haystack,
+    size_t haystack_length,
+    const bitmask_t *needle_mask,
+    bitmask_t zero_mask,
+    bitmask_t found_mask):
+    cdef:
+        bitmask_t R = zero_mask
+        size_t i
+
+    for i in range(haystack_length):
+        # Update the bit array
+        R |= needle_mask[<uint8_t>haystack[i]]
+        R <<= 1
+        R &= zero_mask
```

Comment:
Not really. It is a consequence of where I started. I started with shift-or for single word searches. It was much easier to adapt the already existing code this way. If I had started from scratch shift-and would have been more logic.

What is the best option here. Go for more logial shift-and and then add an explanation, or just explain and forget about making the changes?

Reply:
Don’t make the change. Let’s not perfect be the enemy of the good.

Reply:
I added the explanation. I assume knowledge of shift-or. (Otherwise it will be quite long to explain, and people can look it up).

I am actually quite proud of myself for this "innovation". I could not find this in the literature. Probably someone else thought of it before me though, but it has limited use because of the small size of a machine word. Having said that, with SIMD instructions 128-bit machine words are available on quite a lot of machines (also arm64 by default!). Maybe I should publish something? Though I am 99% sure somebody has already done this, If a biologists can think this up within 2 weeks of learning shift-or. 
At least I can truly call myself a bioinformatician now ;-).

Reply:
Great, nicely written and super helpful! (If there ever is a new Cutadapt paper, this could be a section, could be harder to publish separately.)

Reply:
Found it https://hal.inria.fr/inria-00637227/document
Already published before:

> Algorithm. The idea of our algorithm is to store n fixed-length words into a
machine word, so n matchings can be done simultaneously.

Maybe link the paper in the source code?